src/devices/Gpio/Drivers/VirtualGpioController.cs

```diff
@@ -0,0 +1,318 @@
+            foreach (var pin in pins)
+            {
+                _pins.TryAdd(inc++, pin);
+            }
+        }
+
+        /// <summary>
+        /// Adds a new pin number with an associated <see cref="GpioPin"/>.
+        /// </summary>
+        /// <param name="pinNumber">The pin number.</param>
+        /// <param name="gpioPin">The <see cref="GpioPin"/> to add.</param>
+        public void Add(int pinNumber, GpioPin gpioPin) => _pins.TryAdd(pinNumber, gpioPin);
+
+        /// <summary>
+        /// Adds a new <see cref="GpioPin"/>. The number is done automatically by adding after the last element.
+        /// </summary>
+        /// <param name="gpioPin">The <see cref="GpioPin"/> to add.</param>
+        public void Add(GpioPin gpioPin)
+        {
+            // Find the last number of the pin and assign the new number to it
+            int newPin = _pins.Count > 0 ? _pins.Keys.Max() + 1 : 0;
+            _pins.TryAdd(newPin, gpioPin);
+        }
+
+        /// <inheritdoc/>
+        public override int PinCount => _pins.Count;
+
+        /// <summary>
+        /// Closes an open pin.
+        /// If allowed by the driver, the state of the pin is not changed.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        public new void ClosePin(int pinNumber)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not close pin {pinNumber} because it is not open.");
+            }
+
+            _openPins.TryRemove(pinNumber, out _);
+        }
+
+        /// <summary>
+        /// Disposes this instance and closes all open pins associated with this controller.
+        /// </summary>
+        public new void Dispose()
+        {
+            foreach (int pin in _openPins.Keys)
+            {
+                // The list contains the pin in the current NumberingScheme
+                ClosePin(pin);
+            }
+
+            // We're just emptying the lists
+            _pins.Clear();
+            _openPins.Clear();
+        }
+
+        /// <inheritdoc/>
+        public override PinMode GetPinMode(int pinNumber)
+        {
+            if (!IsPinOpen(pinNumber))
+            {
+                throw new InvalidOperationException($"Can not set a mode to pin {pinNumber} because it is not open.");
+            }
+
+            return _pins[pinNumber].GetPinMode();
+        }
+
+        /// <inheritdoc/>
+        public override bool IsPinModeSupported(int pinNumber, PinMode mode)
+        {
+            return _pins[pinNumber].IsPinModeSupported(mode);
+        }
+
+        /// <summary>
+        /// Checks if a specific pin is open.
+        /// </summary>
+        /// <param name="pinNumber">The pin number in the controller's numbering scheme.</param>
+        /// <returns>The status if the pin is open or closed.</returns>
+        public new bool IsPinOpen(int pinNumber)
```

Comment:
Better make the base method virtual. Otherwise the behavior could be undefined if using the base class.